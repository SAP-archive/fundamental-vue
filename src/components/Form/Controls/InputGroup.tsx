import { Slot, Component, DefaultSlot, Prop, Base } from '@/core';

interface Props {
  before?: string | null;
  after?: string | null;
  afterClass?: string | null;
  compact?: boolean;
}

@Component('InputGroup')
@Slot('before', 'Content to be placed before the input component.')
@Slot('after', 'Content to be placed after the input component.')
@DefaultSlot('The input component placed in the input group.')
export class InputGroup extends Base<Props> {
  @Prop('text/number before the input', { type: String, default: null })
  public before!: string | null;

  @Prop('text/number after the input', { type: String, default: null })
  public after!: string | null;

  @Prop({ type: String, default: null })
  public afterClass!: string | null;

  @Prop('whether input group is compact', { type: Boolean, default: false })
  public compact!: boolean;

  public render() {
    const beforeAddon = this.before || this.$slots.before;
    const afterAddon = this.after || this.$slots.after;
    const afterClass = this.afterClass || '';
    const ws = this.afterClass ? ' ' : '';
    const afterAddonClassName = 'fd-input-group__addon fd-input-group__addon--after' + ws + afterClass;
    return (
      <div staticClass='fd-input-group' class={this.classes}>
        {beforeAddon && <span class='fd-input-group__addon fd-input-group__addon--before'>{beforeAddon}</span>}
        {this.$slots.default}
        {afterAddon && <span class={afterAddonClassName}>{afterAddon}</span>}
      </div>
    );
  }

  private get classes() {
    return {
      'fd-input-group--before': this.before || this.$slots.before,
      'fd-input-group--after': this.after || this.$slots.after,
      'fd-input-group--compact': this.compact,
    };
  }
}
