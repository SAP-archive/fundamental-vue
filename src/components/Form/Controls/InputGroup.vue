<template>
  <div class="fd-input-group" :class="classes">
    <span
      v-if="hasBefore"
      class="fd-input-group__addon fd-input-group__addon--before"
    >
      <slot name="before">{{ before }}</slot>
    </span>
    <slot />
    <span v-if="hasAfter" :class="afterClasses">
      <slot name="after">{{ after }}</slot>
    </span>
  </div>
</template>

<script>
export default {
  name: "FdInputGroup",
  props: {
    before: { type: String, default: null },
    after: { type: String, default: null },
    afterClass: { type: String, default: null },
    compact: { type: Boolean, default: false }
  },
  computed: {
    hasBefore() {
      return this.before != null || this.$slots.before != null;
    },
    hasAfter() {
      return this.after != null || this.$slots.after != null;
    },
    afterClasses() {
      const afterClass = this.afterClass;
      return [
        "fd-input-group__addon",
        "fd-input-group__addon--after",
        this.compact ? "fd-input-group--compact" : "",
        ...(afterClass != null ? [afterClass] : [])
      ];
    },
    classes() {
      return {
        "fd-input-group--before": this.before || this.$slots.before,
        "fd-input-group--after": this.after || this.$slots.after,
        "fd-input-group--compact": this.compact
      };
    }
  }
};
</script>
