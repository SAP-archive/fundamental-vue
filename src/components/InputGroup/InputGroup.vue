<template>
  <div class="fd-input-group" :class="classes">
    <fd-addon-before-provider>
      <slot name="before">
        {{ before }}
      </slot>
    </fd-addon-before-provider>
    <slot name="before-input" />
    <slot name="input" />
    <slot name="after-input" />
    <fd-addon-after-provider>
      <slot name="after">{{ after }}</slot>
    </fd-addon-after-provider>
  </div>
</template>

<script>
import { CompactableContainer } from "./../../mixins";

const CreateAddon = context => ({
  provide() {
    return {
      $_fdInputGroupAddonContext: context
    };
  },
  render() {
    return this.$scopedSlots.default();
  }
});

const FdAddonBeforeProvider = CreateAddon("before");
const FdAddonAfterProvider = CreateAddon("after");

export default {
  name: "FdInputGroup",
  mixins: [CompactableContainer],
  components: { FdAddonBeforeProvider, FdAddonAfterProvider },
  props: {
    before: { type: String, default: null },
    after: { type: String, default: null }
  },
  computed: {
    hasBefore() {
      return this.before != null || this.$slots.before != null;
    },
    hasAfter() {
      return this.after != null || this.$slots.after != null;
    },
    afterClasses() {
      const afterClass = this.afterClass;
      return [
        "fd-input-group__addon",
        "fd-input-group__addon--after",
        this.finalCompact ? "fd-input-group--compact" : "",
        ...(afterClass != null ? [afterClass] : [])
      ];
    },
    classes() {
      return {
        "fd-input-group--before": this.before || this.$slots.before,
        "fd-input-group--after": this.after || this.$slots.after,
        "fd-input-group--compact": this.compact
      };
    }
  }
};
</script>
