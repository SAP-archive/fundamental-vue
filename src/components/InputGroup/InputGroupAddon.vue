<template>
  <span class="fd-input-group__addon" :class="classes"><slot /></span>
</template>

<script>
export default {
  name: "FdInputGroupAddon",
  inject: {
    $_fdInputGroupAddonContext: { default: null }
  },
  provide() {
    return {
      $_fdInputGroupAddon: this
    };
  },
  data() {
    return {
      hasButton: false
    };
  },
  computed: {
    addonContext() {
      return this.$_fdInputGroupAddonContext;
    },
    classes() {
      return {
        "fd-input-group__addon--button": this.hasButton,
        "fd-input-group__addon--before": this.addonContext === "before",
        "fd-input-group__addon--after": this.addonContext === "after"
      };
    }
  }
};
</script>
