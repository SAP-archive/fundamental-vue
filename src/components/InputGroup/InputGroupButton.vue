<template>
  <fd-button styling="light" v-bind="$attrs" v-on="$listeners"><slot /></fd-button>
</template>

<script>
import FdButton from "./../Button/Button.vue";

export default {
  name: "FdInputGroupButton",
  components: { FdButton },
  inject: ["$_fdInputGroupAddon"],
  mounted() {
    this.$_fdInputGroupAddon.hasButton = true;
  }
};
</script>
