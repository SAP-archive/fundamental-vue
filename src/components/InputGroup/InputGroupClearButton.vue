<template>
  <!-- the inline style is a workaround and should be removed -->
  <!-- once it has been fixed in fund. styles. -->
  <button
    class="fd-input-group__button fd-input-group__button--clear"
    style="z-index: 2;"
    aria-label="Clear"
    v-bind="$attrs"
    v-on="$listeners"
  >
    <slot />
  </button>
</template>

<script>
export default {
  name: "FdInputGroupClearButton",
  inject: ["$_fdInputGroupAddon"],
  mounted() {
    this.$_fdInputGroupAddon.hasButton = true;
  }
};
</script>
