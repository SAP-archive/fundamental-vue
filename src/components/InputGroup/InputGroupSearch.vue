<template>
  <fd-input-group>
    <template #input>
      <slot name="input" />
    </template>
    <template #after-input>
      <fd-input-group-addon>
        <slot name="clear" />
      </fd-input-group-addon>
    </template>
  </fd-input-group>
</template>

<script>
import FdInputGroup from "./InputGroup.vue";
import FdInputGroupAddon from "./InputGroupAddon.vue";

export default {
  name: "FdInputGroupSearch",
  components: { FdInputGroup, FdInputGroupAddon }
};
</script>
