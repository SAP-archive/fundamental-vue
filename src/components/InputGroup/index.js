import InputGroup from "./InputGroup.vue";
import InputGroupSearch from "./InputGroupSearch.vue";
import InputGroupAddon from "./InputGroupAddon.vue";
import InputGroupButton from "./InputGroupButton.vue";
import InputGroupClearButton from "./InputGroupClearButton.vue";
import { pluginify } from "./../../util";
export default pluginify(
  InputGroup,
  InputGroupAddon,
  InputGroupButton,
  InputGroupClearButton,
  InputGroupSearch
);
export {
  InputGroup,
  InputGroupAddon,
  InputGroupButton,
  InputGroupClearButton,
  InputGroupSearch
};
