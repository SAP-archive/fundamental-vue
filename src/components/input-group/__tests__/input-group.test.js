import { mount } from '@vue/test-utils'
import InputGroup from '../input-group.vue'
import InputGroupAddon from '../addon.vue'
import InputGroupInput from './../input.vue'

describe('InputGroup', () => {
  const before = '$'
  const after = '€'

  const wrapper = mount({
    components: { InputGroup, InputGroupAddon, InputGroupInput },
    template: `
    <InputGroup>
      <template #before>
        <InputGroupAddon>$</InputGroupAddon>
      </template>
      <template #after>
        <InputGroupAddon>€</InputGroupAddon>
      </template>
      <template #input>
        <InputGroupInput />
      </template>
    </InputGroup>`
  })

  it('renders correctly', () => {
    expect(wrapper.element).toMatchSnapshot()
  })

  // describe("Compact Style", () => {
  //   const wrapper = mount(InputGroup, {
  //     propsData: {
  //       compact: true
  //     },
  //     slots: {
  //       default: {
  //         render(h) {
  //           return h(FdInput);
  //         }
  //       }
  //     }
  //   });

  //   it("renders correctly", () => {
  //     expect(wrapper.element).toMatchSnapshot();
  //   });

  //   it("has compact style", () => {
  //     expect(wrapper.contains(".fd-input-group--compact")).toBe(true);
  //   });

  //   it("does not have before and after", () => {
  //     expect(wrapper.contains(".InputGroup__addon")).toBe(false);
  //   });
  // });
})
