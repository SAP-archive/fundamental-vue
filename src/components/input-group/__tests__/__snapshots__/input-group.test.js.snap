// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InputGroup renders correctly 1`] = `
<div
  class="fd-input-group"
>
  <span
    class="fd-input-group__addon"
  >
    $
  </span>
    
  <input
    class="fd-input fd-input-group__input"
  />
    
  <span
    class="fd-input-group__addon"
  >
    â‚¬
  </span>
</div>
`;
